src/sinks/file.rs

```diff
@@ -0,0 +1,563 @@
+        let sink = PartitionedFileSink::new(
+            Template::from(&self.path[..]),
+            self.close_timeout_secs,
+            self.encoding.clone(),
+        )
+        .stream_ack(acker);
+
+        Ok((Box::new(sink), Box::new(future::ok(()))))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+}
+
+pub type EmbeddedFileSink = Box<dyn Sink<SinkItem = Event, SinkError = ()> + 'static + Send>;
+
+pub struct PartitionedFileSink {
+    path: Template,
+    encoding: Option<Encoding>,
+    close_timeout_secs: Option<u64>,
+
+    partitions: HashMap<Arc<Path>, EmbeddedFileSink>,
+    last_accessed: HashMap<Arc<Path>, Instant>,
+    closing: Vec<EmbeddedFileSink>,
+}
+
+impl PartitionedFileSink {
+    pub fn new(
+        path: Template,
+        close_timeout_secs: Option<u64>,
+        encoding: Option<Encoding>,
+    ) -> Self {
+        PartitionedFileSink {
+            path,
+            encoding,
+            close_timeout_secs,
+            partitions: HashMap::new(),
+            last_accessed: HashMap::new(),
+            closing: Vec::new(),
+        }
+    }
+
+    fn collect_old_files(&mut self) {
+        let mut recently_outdated = Vec::new();
+        if let Some(timeout) = self.close_timeout_secs {
+            self.last_accessed.retain(|path, time| {
+                if time.elapsed().as_secs() > timeout {
+                    debug!(message = "removing file.", file = ?path);
+                    recently_outdated.push(path.clone());
+                    false
+                } else {
+                    true
+                }
+            });
+        }
+
+        let mut recently_outdated = recently_outdated
+            .into_iter()
+            .map(|ref path| self.partitions.remove(path).unwrap())
+            .collect();
+
+        //it is easier to empty a `closing` buffer and then put back all sinks which are not closed yet (see `poll_close`)
+        let mut closing: Vec<EmbeddedFileSink> = self.closing.drain(..).collect();
+        closing.append(&mut recently_outdated);
+
+        for file in closing.into_iter() {
+            self.poll_close(file);
+        }
+    }
+
+    fn poll_close(&mut self, mut file: EmbeddedFileSink) {
+        match file.close() {
+            Err(err) => error!("Error while closing FileSink: {:?}", err),
+            Ok(Async::Ready(())) => debug!("a FileSink closed"),
+            Ok(Async::NotReady) => self.closing.push(file),
+        }
+    }
+
+    fn opened_files(&self) -> Vec<Arc<Path>> {
+        self.partitions.keys().map(|path| path.clone()).collect()
+    }
+}
+
+impl Sink for PartitionedFileSink {
+    type SinkItem = Event;
+    type SinkError = ();
+
+    fn start_send(&mut self, event: Self::SinkItem) -> StartSend<Self::SinkItem, Self::SinkError> {
+        match self.path.render(&event) {
+            Ok(bytes) => match std::str::from_utf8(&bytes) {
```

Comment:
I wonder if it makes more sense to conver to a OsString since that is what PathBuf is based on. What do you think?

Reply:
Interesting point. I would try it, but I think it is necessary to implement better benchmark before serious performance tuning. What is among of my plans since the current one measures only non-partitioned file sink.